accepted/000071-cluster_awareness.md

```diff
@@ -0,0 +1,322 @@
+
+All other actions will be inhibited by the system lock and will require special handling by being implemented in the cluster provider manager level.
+- CaaSP: cordon and drain the node and then reboot. Or, create kured sentinel file https://github.com/SUSE/skuba/blob/master/skuba-update/skuba_update/skuba_update.py#L43
+- SES: enable maintenance mode, reboot, disable maintenance mode (https://documentation.suse.com/ses/6/single-html/ses-admin/#tips-stopping-osd-without-rebalancing)
+- SLE-HA: enable maintenance mode, reboot, disable maintenance mode
+
+#### System groups
+
+When defining a new cluster, Uyuni/SUSE Manager also creates a system group named `<type><name>` in which all nodes of the cluster will be inserted. Every time a node is added or removed from the cluster, Uyuni/SUSE Manager removes the node from the system group.
+
+## Integration requirements from an external product perspective
+
+From the plugin perspective, a product needs to define:
+- a new cluster type in Uyuni/SUSE Manager with a specific configuration (cluster type <-> cluster provider manager)
+- provide a Salt Formula that provisions the management node with cluster management stack
+
+When a new cluster product type is available in Uyuni/SUSE Manager, the association between the cluster type and its cluster provider manager is stored in the database.
+
+Summary of what a product must provide to Uyuni/SUSE Manager:
+
+* CaaSP: Salt-based cluster provider manager ("`caasp`")
+* SES: Salt-based cluster provider manager ("`ses`")
+* SLE-HA/SHAP: Salt-based cluster provider ("`shap`")
+
+It is intended that the cluster provider manager package must at least contain:
+
+- The Salt Formula to deploy the management node
+- All Salt modules and states to implement the cluster actions described above
+
+### Future developments
+
+## Opening Uyuni/SUSE Manager features to cluster objects
+
+Future developments can enrich the implementations described in the RFC by converting the cluster management node into a [Salt proxy minion](https://docs.saltstack.com/en/latest/topics/proxyminion/index.html) using the SSH backend to access all cluster nodes.
+In this way, the cluster object can be represented by a special minion registered to Uyuni/SUSE Manager and the user can issue some operations on the cluster object that will translate the action to all cluster nodes and collect the results. Some examples:
+
+* Assign Salt states
+* Execute Salt execution modules
+* Assign software channels and manage Channel Lifecycle Management
+
+We listed this feature as a future development because:
+* The MVP actions we outlined above have higher priority
+* During our experiments we encountered [this bug](https://github.com/saltstack/salt/issues/53341) with `salt-proxy`, we correctly got past it but we were still unable to register a proxied system to the `salt-master`
+* Some of the features can be already implemented by scripting and using Salt on the cluster.
+
+## SSH keys management
```

Comment:
This will certainly _work_ (and seems to be necessary for `skuba` and probably other tools as well), but it's breaking the encapsulation between the cluster manager and SUSE Manager. We'll need to re-consider the security models of some of those cluster management stacks going forward. It's kind of ridiculous that we'll already have trust established to the physical node via the Salt minion and then still have to add an ssh layer on top just because those tools happen to use ssh as their transport ...

Reply:
Definitely agreed!